### PR TITLE
implemented GetMainContext() for opengl

### PR DESCRIPTION
**Pull request by gaaclarke**
## Description

For Metal, the IOSContext owns the skia context, for OpenGL it was the GPUSurface that owned the context.  I've made the IOSContextGL behave like the IOSContextMetal which allowed us to percolate up a shared interface `GetMainContext()`.  This effectively solves the sharing GPU context problem for OpenGL as well.

## Related Issues

fixes https://github.com/flutter/flutter/issues/72021
https://github.com/flutter/engine/pull/23539

## Tests

I added the following tests:

No new functionality other than changing who has ownership of the skia context.  Existing tests should cover it.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Reviewer Checklist

- [x] I have submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

**Comment by gaaclarke** [comment]
@xster @iskakaushik friendly ping

**Review by xster — COMMENTED** [review]
Doh sorry missed this. LGTM. Could any of this be tested?

**Comment by gaaclarke** [comment]
The new functionality would be covered under an existing test but since we don't run that test on OpenGL with the simulator it won't get exercised.

As for the refactoring of the setup for OpenGL there are golden image tests that run on opengl devices as far as I know.

I'll give unit tests directly on the opengl classes another look later today.

**Review by iskakaushik — APPROVED** [review]

File: shell/gpu/gpu_surface_gl.h [hunk 1]
```diff
@@ -20,6 +20,8 @@ namespace flutter {
 
 class GPUSurfaceGL : public Surface {
  public:
+  static sk_sp<GrDirectContext> MakeGLContext(GPUSurfaceGLDelegate* delegate);
```
Comment (xster) [inline shell/gpu/gpu_surface_gl.h:23]: If possible, can we add some consistency between the metal version and this one? In metal, the IOSContextMetal just decides to make a GrDirectContext the way it wants without asking anyone. Here, it seems like the IOSSurfaceGL asks the GPUSurfaceGL to make one, then pushes it into IOSContextGL which then owns it.

File: shell/platform/darwin/ios/ios_context_gl.mm [hunk 2]
```diff
@@ -24,7 +25,11 @@
   }
 }
 
-IOSContextGL::~IOSContextGL() = default;
+IOSContextGL::~IOSContextGL() {
+  if (main_context_) {
+    main_context_->releaseResourcesAndAbandonContext();
```
Comment (xster) [inline shell/platform/darwin/ios/ios_context_gl.mm:30]: I don't know if it's possible. The various GPUSurfaceGL still has this shared pointer right? Could there be any risks?

File: shell/platform/darwin/ios/ios_surface_gl.mm [hunk 3]
```diff
@@ -38,8 +38,16 @@
 std::unique_ptr<Surface> IOSSurfaceGL::CreateGPUSurface(GrDirectContext* gr_context) {
   if (gr_context) {
     return std::make_unique<GPUSurfaceGL>(sk_ref_sp(gr_context), this, true);
+  } else {
+    IOSContextGL* gl_context = CastToGLContext(GetContext());
+    sk_sp<GrDirectContext> context = gl_context->GetMainContext();
+    if (!context) {
+      context = GPUSurfaceGL::MakeGLContext(this);
+      gl_context->SetMainContext(context);
```
Comment (xster) [inline shell/platform/darwin/ios/ios_surface_gl.mm:46]: Ditto here. Seems like we have 2 timing characteristics. The metal case just makes this GrDirectContext way earlier when creating the shell's on_create_platform_view, whereas this makes it way later in FlutterViewController's viewWillAppear. There could be minutes in between if the engine was prewarmed. 

I'm a bit worried that there might be side-effects involving platform views or memory etc that are hard to reproduce due to the rendering API differences. 

It doesn't have to be in this PR, but I suppose this way works better memory wise. Can we make metal work this way too?

**Pull request merged**